modules/userdirectory-brightspace/src/main/java/org/opencastproject/userdirectory/brightspace/BrightspaceUserProviderInstance.java

```diff
@@ -0,0 +1,274 @@
+      }
+
+      mbs.registerMBean(this, name);
+    } catch (Exception e) {
+      logger.error("Unable to register {} as an mbean: {}", this, e);
+    }
+
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see org.opencastproject.security.api.UserProvider#getName()
+   */
+  @Override
+  public String getName() {
+    return pid;
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see org.opencastproject.security.api.UserProvider#getUsers()
+   */
+  @Override
+  public Iterator<User> getUsers() {
+    return Collections.emptyIterator();
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see org.opencastproject.security.api.UserProvider#loadUser(java.lang.String)
+   */
+  @Override
+  public User loadUser(String userName) {
+    this.loadUserRequests.incrementAndGet();
+    logger.debug("getting user from cache");
+
+    try {
+      Object user = this.cache.getUnchecked(userName);
+      if (user != this.nullToken) {
+        logger.debug("Returning user {} from cache", userName);
+        return (User) user;
+      } else {
+        logger.debug("Returning null user from cache");
+        return null;
+      }
+    } catch (ExecutionError | UncheckedExecutionException ee) {
+      logger.warn("Exception while loading user {}", userName, ee);
+      return null;
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see org.opencastproject.security.api.UserProvider#countUsers()
+   */
+  @Override
+  public long countUsers() {
+    return 0L;
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see org.opencastproject.security.api.UserProvider#getOrganization()
+   */
+  @Override
+  public String getOrganization() {
+    return this.organization.getId();
+  }
+
+  /**
+   * {@inheritDoc}
+   *
+   * @see org.opencastproject.security.api.UserProvider#findUsers(java.lang.String, int, int)
+   */
+  @Override
+  public Iterator<User> findUsers(String query, int offset, int limit) {
+    return Collections.emptyIterator();
+  }
+
+  @Override
+  public void invalidate(String userName) {
+    this.cache.invalidate(userName);
+  }
+
+  @Override
+  public Iterator<Role> getRoles() {
+    return Collections.emptyIterator();
+  }
+
+  @Override
+  public List<Role> getRolesForUser(String username) {
+    User user = this.loadUser(username);
+    if (user != null) {
+      logger.debug("Returning cached role set for {}", username);
+      return new ArrayList<>(user.getRoles());
+    }
+    logger.debug("Return empty role set for {} - not found in Brightspace", username);
+    return Collections.emptyList();
+  }
+
+  @Override
+  public Iterator<Role> findRoles(String query, Target target, int offset, int limit) {
+    return Collections.emptyIterator();
+  }
+
+  private User loadUserFromBrightspace(String username) {
+    if (this.cache == null) {
+      throw new IllegalStateException("The Brightspace user detail service has not yet been configured");
+    } else if (!"admin".equals(username) && !"".equals(username) && !"anonymous".equals(username)) {
```

Comment:
But these are configurable in `custom.properties` so we should never hard-code them:

```
org.opencastproject.security.admin.user=admin
```

Reply:
These providers have all inherited this code from the original Sakai provider, so changes like this should really be made across all of them, or we move the change higher up the stack.

The logic for user providers not handling admin and anonymous is it's not necessary to resolve them further (because admin has all permissions and anonymous is, well, anonymous). So that check could go into the UserDirectoryService or whatever's calling these providers.

Reply:
Then we should fix it in the Sakai user provider as well since that's clearly a bug.

> because admin has all permissions and anonymous is, well, anonymous

The admin user's name in configurable and it could be any other string so there is no guarantee that a user named `admin` has all permissions.

Reply:
Agree with @lkiesow, good catch.  I'm wondering if we should merge this as is, and then fix all the providers with a separate PR after this one and #633 go in.

Reply:
Let's get it in properly. I'll try to provide a reference implementation. Hopefully tomorrow.

Reply:
As promised, I created a reference implementation by fixing the Moodle user provider in pull request #902

Reply:
I have implemented it in the same way for this user provider